File: problems/1301-1400/1329_sort-each-diagonal/solution.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var sortEachDiagonal = function (mat) {
    // Cells on one diagonal share i - j, so sort each diagonal from its
    // top-row / left-column start and write the values back along the walk.
    const m = mat.length;
    const n = mat[0].length;
    const out = Array.from({ length: m }, () => new Array(n).fill(0));
    const starts = [];
    for (let j = 0; j < n; ++j) starts.push([0, j]);
    for (let i = 1; i < m; ++i) starts.push([i, 0]);
    for (const [si, sj] of starts) {
        const diag = [];
        for (let i = si, j = sj; i < m && j < n; ++i, ++j) diag.push(mat[i][j]);
        diag.sort((a, b) => a - b);
        let k = 0;
        for (let i = si, j = sj; i < m && j < n; ++i, ++j) out[i][j] = diag[k++];
    }
    return out;
};

File: problems/1301-1400/1329_sort-each-diagonal/solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {

    public int[][] sortEachDiagonal(int[][] mat) {
        // Cells on one diagonal share i - j, so sort each diagonal from its
        // top-row / left-column start and write the values back along the walk.
        int m = mat.length;
        int n = mat[0].length;
        int[][] out = new int[m][n];
        for (int si = 0; si < m; ++si) {
            scatter(mat, out, si, 0);
        }
        for (int sj = 1; sj < n; ++sj) {
            scatter(mat, out, 0, sj);
        }
        return out;
    }

    private void scatter(int[][] mat, int[][] out, int si, int sj) {
        int m = mat.length;
        int n = mat[0].length;
        List<Integer> diag = new ArrayList<>();
        for (int i = si, j = sj; i < m && j < n; ++i, ++j) {
            diag.add(mat[i][j]);
        }
        Collections.sort(diag);
        int k = 0;
        for (int i = si, j = sj; i < m && j < n; ++i, ++j) {
            out[i][j] = diag.get(k++);
        }
    }
}
